core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java

```diff
@@ -817,19 +823,44 @@ protected boolean shouldCommitDuringPrepare(PrepareCommand command, TxInvocation
              command.isOnePhaseCommit();
    }
 
-   protected final void wrapEntriesForPrepare(TxInvocationContext ctx, PrepareCommand command) throws Throwable {
+   protected final Object wrapEntriesForPrepareAndApply(TxInvocationContext ctx, PrepareCommand command, InvocationSuccessFunction handler) throws Throwable {
       if (!ctx.isOriginLocal() || command.isReplayEntryWrapping()) {
-         for (WriteCommand c : command.getModifications()) {
-            c.setTopologyId(command.getTopologyId());
-            // TODO: we need async invocation since the interceptors may do remote get
-            InvocationStage visitorStage = makeStage(c.acceptVisitor(ctx, entryWrappingVisitor));
-            // Wait for the sub-command to finish. If there was an exception, rethrow it.
-            visitorStage.get();
-
+         return applyModificationsAndThen(ctx, command, command.getModifications(), 0, handler);
+      }
+      return handler.apply(ctx, command, null);
+   }
+
+   private Object applyModificationsAndThen(TxInvocationContext ctx, PrepareCommand command, WriteCommand[] modifications, int startIndex, InvocationSuccessFunction handler) throws Throwable {
+      // We need to execute modifications for the same key sequentially. In theory we could optimize
+      // this loop if there are multiple remote invocations but since remote invocations are rare, we omit this.
+      for (int i = startIndex; i < modifications.length; i++) {
+         WriteCommand c = modifications[i];
+         c.setTopologyId(command.getTopologyId());
+         Object result = c.acceptVisitor(ctx, entryWrappingVisitor);
+
+         InvocationStage visitorStage = null;
+         boolean done = true;
+         if (result instanceof InvocationStage) {
```

Comment:
Such method does not really fit here; I would have to do the cast in order to throw the exception in the else branch. I could use this && helper method to extract && throw, but I don't find that useful.